git-secrets.1
.\" Man page generated from reStructuredText.
.
.TH GIT-SECRETS  "" "" ""
.SH NAME
git-secrets \- Prevents you from committing passwords and other sensitive information to a git repository.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS Contents
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Description\fP
.IP \(bu 2
\fI\%Installing git\-secrets\fP
.INDENT 2.0
.IP \(bu 2
\fI\%*nix (Linux/macOS)\fP
.IP \(bu 2
\fI\%Windows\fP
.IP \(bu 2
\fI\%Homebrew (for macOS users)\fP
.UNINDENT
.IP \(bu 2
\fI\%Advanced configuration\fP
.IP \(bu 2
\fI\%Before making public a repository\fP
.IP \(bu 2
\fI\%Options\fP
.INDENT 2.0
.IP \(bu 2
\fI\%Operation Modes\fP
.IP \(bu 2
\fI\%Options for \fB\-\-install\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-scan\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-list\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-add\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-register\-aws\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-aws\-provider\fP\fP
.IP \(bu 2
\fI\%Options for \fB\-\-add\-provider\fP\fP
.UNINDENT
.IP \(bu 2
\fI\%Defining prohibited patterns\fP
.IP \(bu 2
\fI\%Ignoring false positives\fP
.IP \(bu 2
\fI\%Secret providers\fP
.IP \(bu 2
\fI\%Example walkthrough\fP
.IP \(bu 2
\fI\%Skipping validation\fP
.IP \(bu 2
\fI\%About\fP
.UNINDENT
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan [\-r|\-\-recursive] [\-\-cached] [\-\-no\-index] [\-\-untracked] [<files>...]
git secrets \-\-scan\-history
git secrets \-\-install [\-f|\-\-force] [<target\-directory>]
git secrets \-\-list [\-\-global]
git secrets \-\-add [\-a|\-\-allowed] [\-l|\-\-literal] [\-\-global] <pattern>
git secrets \-\-add\-provider [\-\-global] <command> [arguments...]
git secrets \-\-register\-aws [\-\-global]
git secrets \-\-aws\-provider [<credentials\-file>]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
\fBgit\-secrets\fP scans commits, commit messages, and \fB\-\-no\-ff\fP merges to
prevent adding secrets into your git repositories. If a commit,
commit message, or any commit in a \fB\-\-no\-ff\fP merge history matches one of
your configured prohibited regular expression patterns, then the commit is
rejected.
.SH INSTALLING GIT-SECRETS
.sp
\fBgit\-secrets\fP must be placed somewhere in your PATH so that it is picked up
by \fBgit\fP when running \fBgit secrets\fP\&.
.SS *nix (Linux/macOS)
.IP "System Message: WARNING/2 (README.rst:, line 43)"
Title underline too short.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\e*nix (Linux/macOS)
~~~~~~~~~~~~~~~~~
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can use the \fBinstall\fP target of the provided Makefile to install \fBgit secrets\fP and the man page.
You can customize the install path using the PREFIX and MANPREFIX variables.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
make install
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Windows
.sp
Run the provided \fBinstall.ps1\fP powershell script. This will copy the needed files
to an installation directory (\fB%USERPROFILE%/.git\-secrets\fP by default) and add
the directory to the current user \fBPATH\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
PS > ./install.ps1
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Homebrew (for macOS users)
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
brew install git\-secrets
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBWARNING:\fP
.INDENT 0.0
.INDENT 3.5
You\(aqre not done yet! You MUST install the git hooks for every repo that
you wish to use with \fBgit secrets \-\-install\fP\&.
.UNINDENT
.UNINDENT
.sp
Here\(aqs a quick example of how to ensure a git repository is scanned for secrets
on each commit:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cd /path/to/my/repo
git secrets \-\-install
git secrets \-\-register\-aws
.ft P
.fi
.UNINDENT
.UNINDENT
.SH ADVANCED CONFIGURATION
.sp
Add a configuration template if you want to add hooks to all repositories you
initialize or clone in the future.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-register\-aws \-\-global
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Add hooks to all your local repositories.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-install ~/.git\-templates/git\-secrets
git config \-\-global init.templateDir ~/.git\-templates/git\-secrets
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Add custom providers to scan for security credentials.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add\-provider \-\- cat /path/to/secret/file/patterns
.ft P
.fi
.UNINDENT
.UNINDENT
.SH BEFORE MAKING PUBLIC A REPOSITORY
.sp
With git\-secrets is also possible to scan a repository including all revisions:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan\-history
.ft P
.fi
.UNINDENT
.UNINDENT
.SH OPTIONS
.SS Operation Modes
.sp
Each of these options must appear first on the command line.
.INDENT 0.0
.TP
.B \fB\-\-install\fP
Installs git hooks for a repository. Once the hooks are installed for a git
repository, commits and non\-fast\-forward merges for that repository will be prevented
from committing secrets.
.TP
.B \fB\-\-scan\fP
Scans one or more files for secrets. When a file contains a secret, the
matched text from the file being scanned will be written to stdout and the
script will exit with a non\-zero status. Each matched line will be written with
the name of the file that matched, a colon, the line number that matched,
a colon, and then the line of text that matched. If no files are provided,
all files returned by \fBgit ls\-files\fP are scanned.
.TP
.B \fB\-\-scan\-history\fP
Scans repository including all revisions. When a file contains a secret, the
matched text from the file being scanned will be written to stdout and the
script will exit with a non\-zero status. Each matched line will be written with
the name of the file that matched, a colon, the line number that matched,
a colon, and then the line of text that matched.
.TP
.B \fB\-\-list\fP
Lists the \fBgit\-secrets\fP configuration for the current repo or in the global
git config.
.TP
.B \fB\-\-add\fP
Adds a prohibited or allowed pattern.
.TP
.B \fB\-\-add\-provider\fP
Registers a secret provider. Secret providers are executables that when
invoked output prohibited patterns that \fBgit\-secrets\fP should treat as
prohibited.
.TP
.B \fB\-\-register\-aws\fP
Adds common AWS patterns to the git config and ensures that keys present
in \fB~/.aws/credentials\fP are not found in any commit. The following
checks are added:
.INDENT 7.0
.IP \(bu 2
AWS Access Key IDs via \fB(A3T[A\-Z0\-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A\-Z0\-9]{16}\fP
.IP \(bu 2
Amazon Bedrock API keys. Long\-lived via \fBABSK[A-Za-z0-9+/]{109,}=*\fP and short\-lived via \fBbedrock\-api\-key\-YmVkcm9jay5hbWF6b25hd3MuY29t[a-zA-Z0-9]{1,}\fP
.IP \(bu 2
AWS Secret Access Key assignments via ":" or "=" surrounded by optional
quotes
.IP \(bu 2
AWS account ID assignments via ":" or "=" surrounded by optional quotes
.IP \(bu 2
Allowed patterns for example AWS keys (\fBAKIAIOSFODNN7EXAMPLE\fP and
\fBwJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\fP)
.IP \(bu 2
Known credentials from \fB~/.aws/credentials\fP
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
While the patterns registered by this command should catch most
instances of AWS credentials, these patterns are \fBnot\fP guaranteed to
catch them \fBall\fP\&. \fBgit\-secrets\fP should be used as an extra means of
insurance \-\- you still need to do your due diligence to ensure that you
do not commit credentials to a repository.
.UNINDENT
.UNINDENT
.TP
.B \fB\-\-aws\-provider\fP
Secret provider that outputs credentials found in an INI file. You can
optionally provide the path to an INI file.
.UNINDENT
.SS Options for \fB\-\-install\fP
.INDENT 0.0
.TP
.B \fB\-f, \-\-force\fP
Overwrites existing hooks if present.
.TP
.B \fB<target\-directory>\fP
When provided, installs git hooks to the given directory. The current
directory is assumed if \fB<target\-directory>\fP is not provided.
.sp
If the provided \fB<target\-directory>\fP is not in a git repository, the
directory will be created and hooks will be placed in
\fB<target\-directory>/hooks\fP\&. This can be useful for creating git template
directories using with \fBgit init \-\-template <target\-directory>\fP\&.
.sp
You can run \fBgit init\fP on a repository that has already been initialized.
From the \fI\%git init documentation\fP:
.INDENT 7.0
.INDENT 3.5
From the git documentation: Running \fBgit init\fP in an existing repository
is safe. It will not overwrite things that are already there. The
primary reason for rerunning \fBgit init\fP is to pick up newly added
templates (or to move the repository to another place if
\fB\-\-separate\-git\-dir\fP is given).
.UNINDENT
.UNINDENT
.sp
The following git hooks are installed:
.INDENT 7.0
.IP 1. 3
\fBpre\-commit\fP: Used to check if any of the files changed in the commit
use prohibited patterns.
.IP 2. 3
\fBcommit\-msg\fP: Used to determine if a commit message contains a
prohibited patterns.
.IP 3. 3
\fBprepare\-commit\-msg\fP: Used to determine if a merge commit will
introduce a history that contains a prohibited pattern at any point.
Please note that this hook is only invoked for non fast\-forward merges.
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
Git only allows a single script to be executed per hook. If the
repository contains Debian\-style subdirectories like \fBpre\-commit.d\fP
and \fBcommit\-msg.d\fP, then the git hooks will be installed into these
directories, which assumes that you\(aqve configured the corresponding
hooks to execute all of the scripts found in these directories. If
these git subdirectories are not present, then the git hooks will be
installed to the git repo\(aqs \fB\&.git/hooks\fP directory.
.UNINDENT
.UNINDENT
.UNINDENT
.SS Examples
.sp
Install git hooks to the current directory:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cd /path/to/my/repository
git secrets \-\-install
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Install git hooks to a repository other than the current directory:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-install /path/to/my/repository
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Create a git template that has \fBgit\-secrets\fP installed, and then copy that
template into a git repository:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-install ~/.git\-templates/git\-secrets
git init \-\-template ~/.git\-templates/git\-secrets
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Overwrite existing hooks if present:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-install \-f
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Options for \fB\-\-scan\fP
.INDENT 0.0
.TP
.B \fB\-r, \-\-recursive\fP
Scans the given files recursively. If a directory is encountered, the
directory will be scanned. If \fB\-r\fP is not provided, directories will be
ignored.
.sp
\fB\-r\fP cannot be used alongside \fB\-\-cached\fP, \fB\-\-no\-index\fP, or
\fB\-\-untracked\fP\&.
.TP
.B \fB\-\-cached\fP
Searches blobs registered in the index file.
.TP
.B \fB\-\-no\-index\fP
Searches files in the current directory that is not managed by git.
.TP
.B \fB\-\-untracked\fP
In addition to searching in the tracked files in the working tree,
\fB\-\-scan\fP also in untracked files.
.TP
.B \fB<files>...\fP
The path to one or more files on disk to scan for secrets.
.sp
If no files are provided, all files returned by \fBgit ls\-files\fP are
scanned.
.UNINDENT
.SS Examples
.sp
Scan all files in the repo:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Scans a single file for secrets:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan /path/to/file
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Scans a directory recursively for secrets:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan \-r /path/to/directory
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Scans multiple files for secrets:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan /path/to/file /path/to/other/file
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can scan by globbing:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-scan /path/to/directory/*
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Scan from stdin:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo \(aqhello!\(aq | git secrets \-\-scan \-
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Options for \fB\-\-list\fP
.INDENT 0.0
.TP
.B \fB\-\-global\fP
Lists only git\-secrets configuration in the global git config.
.UNINDENT
.SS Options for \fB\-\-add\fP
.INDENT 0.0
.TP
.B \fB\-\-global\fP
Adds patterns to the global git config
.TP
.B \fB\-l, \-\-literal\fP
Escapes special regular expression characters in the provided pattern so
that the pattern is searched for literally.
.TP
.B \fB\-a, \-\-allowed\fP
Mark the pattern as allowed instead of prohibited. Allowed patterns are
used to filter out false positives.
.TP
.B \fB<pattern>\fP
The regex pattern to search.
.UNINDENT
.SS Examples
.sp
Adds a prohibited pattern to the current repo:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \(aq[A\-Z0\-9]{20}\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Adds a prohibited pattern to the global git config:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \-\-global \(aq[A\-Z0\-9]{20}\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Adds a string that is scanned for literally (\fB+\fP is escaped):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \-\-literal \(aqfoo+bar\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Add an allowed pattern:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \-a \(aqallowed pattern\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Options for \fB\-\-register\-aws\fP
.INDENT 0.0
.TP
.B \fB\-\-global\fP
Adds AWS specific configuration variables to the global git config.
.UNINDENT
.SS Options for \fB\-\-aws\-provider\fP
.INDENT 0.0
.TP
.B \fB[<credentials\-file>]\fP
If provided, specifies the custom path to an INI file to scan. If not
provided, \fB~/.aws/credentials\fP is assumed.
.UNINDENT
.SS Options for \fB\-\-add\-provider\fP
.INDENT 0.0
.TP
.B \fB\-\-global\fP
Adds the provider to the global git config.
.TP
.B \fB<command>\fP
Provider command to invoke. When invoked the command is expected to write
prohibited patterns separated by new lines to stdout. Any extra arguments
provided are passed on to the command.
.UNINDENT
.SS Examples
.sp
Registers a secret provider with arguments:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add\-provider \-\- git secrets \-\-aws\-provider
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Cats secrets out of a file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add\-provider \-\- cat /path/to/secret/file/patterns
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DEFINING PROHIBITED PATTERNS
.sp
\fBegrep\fP\-compatible regular expressions are used to determine if a commit or
commit message contains any prohibited patterns. These regular expressions are
defined using the \fBgit config\fP command. It is important to note that
different systems use different versions of egrep. For example, when running on
macOS, you will use a different version of \fBegrep\fP than when running on something
like Ubuntu (BSD vs GNU).
.sp
You can add prohibited regular expression patterns to your git config using
\fBgit secrets \-\-add <pattern>\fP\&.
.SH IGNORING FALSE POSITIVES
.sp
Sometimes a regular expression might match false positives. For example, git
commit SHAs look a lot like AWS access keys. You can specify many different
regular expression patterns as false positives using the following command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \-\-allowed \(aqmy regex pattern\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can also add regular expressions patterns to filter false positives to a
\fB\&.gitallowed\fP file located in the repository\(aqs root directory. Lines starting
with \fB#\fP are skipped (comment line) and empty lines are also skipped.
.sp
First, git\-secrets will extract all lines from a file that contain a prohibited
match. Included in the matched results will be the full path to the name of
the file that was matched, followed by \(aq:\(aq, followed by the line number that was
matched, followed by the entire line from the file that was matched by a secret
pattern. Then, if you\(aqve defined allowed regular expressions, git\-secrets will
check to see if all of the matched lines match at least one of your registered
allowed regular expressions. If all of the lines that were flagged as secret
are canceled out by an allowed match, then the subject text does not contain
any secrets. If any of the matched lines are not matched by an allowed regular
expression, then git\-secrets will fail the commit/merge/message.
.sp
\fBIMPORTANT:\fP
.INDENT 0.0
.INDENT 3.5
Just as it is a bad practice to add prohibited patterns that are too
greedy, it is also a bad practice to add allowed patterns that are too
forgiving. Be sure to test out your patterns using ad\-hoc calls to
\fBgit secrets \-\-scan $filename\fP to ensure they are working as intended.
.UNINDENT
.UNINDENT
.SH SECRET PROVIDERS
.sp
Sometimes you want to check for an exact pattern match against a set of known
secrets. For example, you might want to ensure that no credentials present in
\fB~/.aws/credentials\fP ever show up in a commit. In these cases, it\(aqs better to
leave these secrets in one location rather than spread them out across git
repositories in git configs. You can use "secret providers" to fetch these
types of credentials. A secret provider is an executable that when invoked
outputs prohibited patterns separated by new lines.
.sp
You can add secret providers using the \fB\-\-add\-provider\fP command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add\-provider \-\- git secrets \-\-aws\-provider
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Notice the use of \fB\-\-\fP\&. This ensures that any arguments associated with the
provider are passed to the provider each time it is invoked when scanning
for secrets.
.SH EXAMPLE WALKTHROUGH
.sp
Let\(aqs take a look at an example. Given the following subject text (stored in
\fB/tmp/example\fP):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
This is a test!
password=ex@mplepassword
password=******
More test...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
And the following registered patterns:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \(aqpassword\es*=\es*.+\(aq
git secrets \-\-add \-\-allowed \-\-literal \(aqex@mplepassword\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Running \fBgit secrets \-\-scan /tmp/example\fP, the result will
result in the following error output:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
/tmp/example:3:password=******

[ERROR] Matched prohibited pattern

Possible mitigations:
\- Mark false positives as allowed using: git config \-\-add secrets.allowed ...
\- List your configured patterns: git config \-\-get\-all secrets.patterns
\- List your configured allowed patterns: git config \-\-get\-all secrets.allowed
\- Use \-\-no\-verify if this is a one\-time false positive
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Breaking this down, the prohibited pattern value of \fBpassword\es*=\es*.+\fP will
match the following lines:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
/tmp/example:2:password=ex@mplepassword
/tmp/example:3:password=******
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\&...But the first match will be filtered out due to the fact that it matches the
allowed regular expression of \fBex@mplepassword\fP\&. Because there is still a
remaining line that did not match, it is considered a secret.
.sp
Because that matching lines are placed on lines that start with the filename
and line number (e.g., \fB/tmp/example:3:...\fP), you can create allowed
patterns that take filenames and line numbers into account in the regular
expression. For example, you could whitelist an entire file using something
like:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
git secrets \-\-add \-\-allowed \(aq/tmp/example:.*\(aq
git secrets \-\-scan /tmp/example && echo $?
# Outputs: 0
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Alternatively, you could allow a specific line number of a file if that
line is unlikely to change using something like the following:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C

CHANGELOG.md
# CHANGELOG

## 1.3.0 - 2019-02-10

- Empty provider output is now excluded
  <https://github.com/awslabs/git-secrets/issues/34>
- Spaces are now supported in git exec path, making more Windows
  paths execute properly.
- Patterns with newlines and carriage returns are now loaded properly.
- Patterns that contain only "\n" are now ignored.
- Various Bash 4 fixes <https://github.com/awslabs/git-secrets/issues/66>.
- Make IAM key scanning much more targeted.

## 1.2.1 - 2016-06-27

- Fixed an issue where secret provider commands were causing "command not
  found" errors due to a previously set IFS variable.
  <https://github.com/awslabs/git-secrets/pull/30>

## 1.2.0 - 2016-05-23

- Fixed an issue where spaces files with spaces in their names were not being
  properly scanned in the pre-commit hook.
- Now ignoring empty lines and comments (e.g., `#`) in the .gitallowed file.
- Fixed an issue where numbers were being compared to strings causing failures
  on some platforms.

## 1.1.0 - 2016-04-06

- Bug fix: the pre-commit hook previously only scanned the working directory
  rather than staged files. This release updates the pre-commit hook to instead
  scan staged files so that git-secrets will detect violations if the working
  directory drifts from the staging directory.
- Added the `--scan-history` subcommand so that you can scan your entire
  git history for violations.
- Added the ability to filter false positives by using a .gitallowed file.
- Added support for `--cached`, `--no-index`, and `--untracked` to the `--scan`
  subcommand.

## 1.0.1 - 2016-01-11

- Now works correctly with filenames in a repository that contain spaces when
  executing `git secrets --scan` with no provided filename (via `git grep`).
- Now works with git repositories with hundreds of thousands of files when
  using `git secrets --scan` with no provided filename (via `git grep`).

## 1.0.0 - 2015-12-10

- Initial release of `git-secrets`.


git-secrets.sh
#!/usr/bin/env bash
declare -r VERSION="0.0.1"
declare PATTERNS=""
[[ -t 0 && -t 1 ]] && declare -r IS_ATTY=1 || declare -r IS_ATTY=0

#######################################################################
# Help and error messages
#######################################################################

color() {
  [ $IS_ATTY -eq 1 ] && echo "$(tput setaf $1)${@:2}$(tput sgr 0)" || echo $2
}

die() { color 1 "$@" >&2; exit 1; }

version() {
  echo "git-secrets ${VERSION}"
}

usage() {
  version
  echo ""
  echo "Usage: git secrets <command> [<options>] [<args>]"
  echo
  echo "Options:"
  echo "  -h  Displays this message."
  echo "  -v  Displays the version number."
  echo
  echo "Commands:"
  echo "  scan     Scans the passed in file name for prohibited patterns."
  echo "  install  Installs git hooks for a repository."
  exit 0
}

scan_usage() {
  echo "Usage: git secrets scan [<options>] <filename>"
  echo
  echo "Scans the given file (FILENAME) for a list of prohibited patterns"
  echo "found by calling git config --get-all secrets.pattern. If any of"
  echo "these prohibited patterns are found found in the file, then the match"
  echo "is printed and the script fails."
  echo
  echo "Options:"
  echo "  -h  Displays this message."
  echo
  echo "Arguments:"
  echo "  filename  The path to a file on disk to scan. Pass - for stdin."
  exit 0
}

install_usage() {
  echo "Usage: git secrets install [<options>] <repo>"
  echo
  echo "Installs pre-commit and commit-msg hooks for the given repository."
  echo "If no repository argument is provided, then the command attempts"
  echo "to install hooks to a git repository in the working directory."
  echo
  echo "Git only allows a single script to be executed per hook. If the"
  echo "repository contains Debian style subdirectories like pre-commit.d"
  echo "and commit-msg.d, then the git hooks will be installed into these"
  echo "directories, which assumes that you've configured the corresponding"
  echo "hooks to execute all of the scripts found in these directories."
  echo
  echo "Options:"
  echo "  -h  Displays this message."
  echo
  echo "Arguments:"
  echo "  repo  The path to a git repository."
  exit 0
}

# Prints a warning that the commit contains a prohibited pattern.
# Arguments: $1: The first line of text to print.
prohibited_warning() {
  echo
  echo "Error: Prohibited pattern match"
  echo "==============================="
  echo
  echo "$1"
  echo
  echo "You can view the prohibited pattern matches in the above output."
  echo "Please remove these patterns, rebase, and resubmit the patch."
}

#######################################################################
# Helper functions
#######################################################################

load_secret_patterns() {
  PATTERNS="$(git config --get-all secrets.pattern)"
  # Warn if no patterns are configured.
  if [ -z "$PATTERNS" ]; then
    color 3 "No prohibited patterns have been defined" 1>&2
    cat << EOF 1>&2
========================================
You can add prohibited patterns by editing your .git/config file or by using
the following command for each prohibited pattern you wish to add:

    git config --add secrets.pattern <regex-pattern>

You can list all of the configured prohibited patterns by running the
following command:

    git config --get-all secrets.pattern
EOF
  fi
}

git_repo_root() {
  git status > /dev/null 2>&1 && git rev-parse --show-toplevel
}

check_pattern() {
  local pattern="$1" filename="$2" grep_cmd=$(git config --get secrets.grep)
  # Ensure the grep command is valid and executable.
  grep_cmd=$(which "${grep_cmd:-egrep}")
  [ ! -x "$grep_cmd" ] && die "Invalid secrets.grep command: ${grep_cmd}"
  GREP_OPTIONS='' $grep_cmd --colour -nw -H -e "${pattern}" "${filename}"
}

validate_filename() {
  [ -z "$1" ] && die "Empty or missing filename argument."
  [ ! -f "$1" ] && die "File not found: $1"
}

#######################################################################
# Git hook implementation functions
#######################################################################

prompt_for_input() {
  local pattern=$1 file=$2
  local -r affirmative="Yes. This is a false positive, allow it and continue."
  local -r negative="No. This is a secret. Do not commit it -- fail."
  echo "^^^ Matched prohibited pattern in ${file}: ${pattern}"
  echo "Do you wish to commit this anyway?"
  select yn in "${negative}" "${affirmative}"; do
    case $yn in
      [nN]* ) die "Found a secret in ${file}";;
      [Y]* ) break;;
    esac
  done
}

# Scans a file for prohibited patterns.
scan() {
  local -r filename="$1"
  load_secret_patterns
  # Validate the filename only if it is not stdin ("-")
  [ "${filename}" != "-" ] && validate_filename "${filename}"
  if [ ! -z "$PATTERNS" ]; then
    if [ $IS_ATTY -ne 1 ]; then
      # Check will all patterns at once
      check_pattern "$PATTERNS" "${filename}" && die "Found secrets"
    else
      # Check each pattern individually for user prompts.
      for pattern in $PATTERNS; do
        if check_pattern "${pattern}" "${filename}"; then
          prompt_for_input "${pattern}" "${filename}"
        fi
      done
    fi
  fi
}

# Scans a commit message for prohibited patterns.
commit_msg_hook() {
  local -r commit_msg_file="$1"
  git secrets scan "${commit_msg_file}" && exit 0
  prohibited_warning "Your commit message contains a prohibited pattern."
  exit 1
}

# NOTE: This is based on git's pre-commit.sample script.
# Allows the hook to reject commits for brand new repos.
determine_rev_to_diff() {
  [ git rev-parse --verify HEAD >/dev/null 2>&1 ] \
    && "HEAD" || echo "4b825dc642cb6eb9a060e54bf8d69288fbee4904"
}

# Scans a commit for prohibited patterns.
pre_commit_hook() {
  local file found_match=0 against=$(determine_rev_to_diff)
  local changes=$(git diff-index --name-status --cached $against -- | cut -c3-)

  for file in $changes; do
    git secrets scan "${file}" || found_match=1
  done

  if [ "$found_match" -eq 1 ]; then
    prohibited_warning "Your commit contains a prohibited pattern."
    exit 1
  fi
}

# Determines if merging in a commit will introduce tainted history.
prepare_commit_msg_hook() {
  case "$2,$3" in
    merge,)
      local git_head=$(env | grep GITHEAD)  # e.g. GITHEAD_<sha>=release/1.43
      local sha="${git_head##*=}"           # Get just the SHA
      local branch=$(git symbolic-ref HEAD) # e.g. refs/heads/master
      local dest="${branch#refs/heads/}"    # cut out "refs/heads"
      echo "Checking if merging ${sha} into ${dest} adds prohibited history"
      git log "${dest}".."${sha}" -p | git secrets scan -
      ;;
    *) ;;
  esac
}

#######################################################################
# Git hook installation functions
#######################################################################

# Determines the approriate path for a hook to be installed
# This function respects any found $hook.d directories.
determine_hook_path() {
  local -r hook="$1" path="$2"
  local dest="${path}/.git/hooks/${hook}"
  local -r debian_dir="${path}/.git/hooks/${hook}.d"
  [ -d "${debian_dir}" ] && dest="${debian_dir}/git-secrets"
  [ -f "$dest" ] && color 3 "Overwriting $dest" 1>&2
  echo "$dest"
}

install_hook() {
  local -r dest="$1" name="$2" cmd="$3"
  [ -z "${dest}" ] && die "Expects the path to a file"
  [ -d "$(dirname ${dest})" ] || die "Directory not found: ${dest}"
  echo "#!/usr/bin/env bash" > "${dest}"
  echo "git secrets ${cmd} \"\$@\"" >> "${dest}"
  chmod +x "${dest}"
  color 2 "Installed ${name} hook to ${dest}"
}

install_hooks() {
  local dest git_repo="$1"
  dest=$(determine_hook_path "commit-msg" "${git_repo}")
  install_hook "${dest}" "commit-msg" "commit_msg_hook"
  dest=$(determine_hook_path "pre-commit" "${git_repo}")
  install_hook "${dest}" "pre-commit" "pre_commit_hook"
  dest=$(determine_hook_path "prepare-commit-msg" "${git_repo}")
  install_hook "${dest}" "prepare-commit-msg" "prepare_commit_msg_hook"
}

#######################################################################
# Dispatch to the appropriate functions
#######################################################################

main() {
  case "$cmd" in
    -h|--help|'') usage ;;
    -v|--version) version && exit 0 ;;
    commit_msg_hook|pre_commit_hook|prepare_commit_msg_hook) $cmd "$@" ;;
    scan)
      [ "$1" == "-h" ] && scan_usage
      scan "$@"
      ;;
    install)
      [ "$1" == "-h" ] && install_usage
      local git_repo_dir="$1"
      [ -z "${git_repo_dir}" ] && git_repo_dir=$(git_repo_root)
      install_hooks "$git_repo_dir"
      ;;
    *) die "Unknown command: $cmd" ;;
  esac
}

cmd=$1
shift
main "$@"
